### PR TITLE
Implement configurable tab stop width

### LINKED ISSUE #488: Option for configurable tab width?
`xterm.js` inherits from `term.js` the [default tab width of 8 characters](https://github.com/sourcelair/xterm.js/blob/ed13218f9098a997283d996e7570d5cdb24ed87f/src/xterm.js#L3146):

```js
for (; i < this.cols; i += 8) {
  this.tabs[i] = true;
}
```

I think it would be very useful, especially for space-constrained web applications, to be able to set this width to a lower setting.

### PR DESCRIPTION
**Pull request by parisk**
This PR introduces configurable tab stop width.

## Current issues
- [x] No way to call `term.setupStops()` after setting the option, without a hack into the private API. A private `_optionChanged` event could definitely solve this. Ideas @Tyriar?
- [ ] There is no easy way to re-render already printed tab stops. In order to achieve this we will have to separate the data from the presentation layer and re-render the whole thing, which is a big issue.

Documentation PR: xtermjs/xtermjs.org#13

Closes #488.

### CODE REVIEW

**Comment by Tyriar** [comment]
What does `setupStops` do exactly? We can call it from within `getOption` like is done to fix `scrollback`?

**Comment by parisk** [comment]
`setupStops` defines the position in the terminal where tab stops are located.

I added it into `setOption` just like we do with `cursorBlink`, although I wasn't quite fond of the idea at first.

**Comment by parisk** [comment]
@Tyriar I rebased with master and resolved conflicts. This is ready for review.

File: demo/index.html [hunk 1]
```diff
@@ -17,10 +17,13 @@
         <div>
           <h2>Options</h2>
           <p>
-            <label><input type="checkbox" id="option-cursor-blink"> cursorBlink</label>
+            <label>cursorBlink <input type="checkbox" id="option-cursor-blink"></label>
```
Comment (Tyriar) [inline demo/index.html]: Normally checkboxes are at the start of the line?

File: demo/index.html [hunk 2]
```diff
@@ -17,10 +17,13 @@
...
           </p>
           <p>
-            <label>Scrollback <input type="number" id="option-scrollback" value="1000" /></label>
+            <label>scrollback <input type="number" id="option-scrollback" value="1000" /></label>
+          </p>
+          <p>
+            <label>tabStopWidth <input type="number" id="option-tabstopwidth" value="4" /></label>
```
Comment (Tyriar) [inline demo/index.html]: value=8 as the default?

File: demo/main.js [hunk 3]
```diff
@@ -32,7 +33,10 @@ optionElements.cursorBlink.addEventListener('change', function () {
   term.setOption('cursorBlink', optionElements.cursorBlink.checked);
 });
 optionElements.scrollback.addEventListener('change', function () {
-  terminal.setOption('scrollback', parseInt(optionElements.scrollback.value, 10));
+  term.setOption('scrollback', parseInt(optionElements.scrollback.value, 10));
+});
+optionElements.tabstopwidth.addEventListener('change', function () {
+  term.setOption('tabStopWidth', parseInt(optionElements.tabstopwidth.value));
```
Comment (Tyriar) [inline demo/main.js]: Add radix of 10 as second arg as it's a mandatory arg https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

File: demo/main.js [hunk 4]
```diff
@@ -44,7 +48,8 @@ function createTerminal() {
   }
   term = new Terminal({
     cursorBlink: optionElements.cursorBlink.checked,
-    scrollback: parseInt(optionElements.scrollback.value, 10)
+    scrollback: parseInt(optionElements.scrollback.value, 10),
+    tabStopWidth: parseInt(optionElements.tabstopwidth.value)
```
Comment (Tyriar) [inline demo/main.js]: Radix arg

File: src/xterm.js [hunk 5]
```diff
@@ -433,6 +434,7 @@ Terminal.prototype.setOption = function(key, value) {
       this.element.classList.toggle(`xterm-cursor-style-underline`, value === 'underline');
       this.element.classList.toggle(`xterm-cursor-style-bar`, value === 'bar');
       break;
+    case 'tabStopWidth': this.setupStops(); break;
```
Comment (Tyriar) [inline src/xterm.js]: Eventually we could move this to event listeners which would probably be a better structure.

**Comment by Tyriar** [comment]
Changing tab will likely break `ls` output? How does `ls` fetch the terminal's tab size? Related: https://github.com/sourcelair/xterm.js/issues/259

File: demo/index.html (hunk 1 above)
Comment (parisk) [inline demo/index.html]: Did this to keep uniformity with the rest. Will revert.

File: demo/index.html (hunk 2 above)
Comment (parisk) [inline demo/index.html]: Nah, forgot it from testing.

File: src/xterm.js (hunk 5 above)
Comment (parisk) [inline src/xterm.js]: Yep. It does the job for now, so let's keep this for another PR.

**Comment by Tyriar** [comment]
Looking at the tab stuff for https://github.com/sourcelair/xterm.js/issues/259, I don't think re-rendering would actually be desirable. There is the issue of how programs get the tab size though.

**Comment by parisk** [comment]
Everything seems to be working fine. I think that re-rendering is an open question and that we can revisit it in the future.

`ls` looks to be working just fine. I am pretty sure though that it is using spaces and not tabs.

## Screnshot
image

**Comment by parisk** [comment]
I also fixed comments in demo, so @Tyriar feel free to take another look.

**Comment by Tyriar** [comment]
It was certainly tabs on my Macbook, I'll grab your branch and verify.

**Comment by Tyriar** [comment]
This is what I see on a dir with a bunch of files of varying lengths. The first is default (8), the second is 4 via `setOption`:

screen shot 2017-01-20 at 2 49 47 pm

**Comment by Tyriar** [comment]
This also happens on Terminal.app, maybe this is just a price you pay?

**Comment by Tyriar** [comment]
You can fix this on linux by doing something like:

```
alias ls=ls --tabsize=4
```

I guess that's just how it is 😄

**Comment by parisk** [comment]
Well, while this view sucks, technically there is nothing wrong about it. Take a close look and you will see that each file name is being printed at the first available tab stop.

The tab stop width though is not available somehow to `ls` in order to take care of properly aligning everything. I guess this is where you have to use `ls --tabsize=4`.

My opinion is that this is the price you have to pay eventually for such a feature and that you should be careful with using it.

I guess it has it's use cases though.

**Comment by Tyriar** [comment]
Yeah, this is the behavior of other terminal emulators that set via calling `tabs` so :+1:

**Comment by vincentwoo** [comment]
Sorry was late to the party. I agree that tab stop width changes should *not* rerender previous content. Otherwise, I have no comments!
